ft_statistics_mvpa.m

```diff
@@ -0,0 +1,231 @@
+% different datasets (one for training the classifier, the other one for
+% testing). This could be realised eg using extra fields in cfg such as
+% cfg.X2 for dataset 2 and cfg.design2 for the second design matrix
+%
+% Returns:
+%   stat        = struct with results. the .metric field contains the
+%                   requested metrics
+% 
+% (c) matthias treder
+
+% do a sanity check on the input data
+assert(isnumeric(dat),    'this function requires numeric data as input, you probably want to use FT_TIMELOCKSTATISTICS, FT_FREQSTATISTICS or FT_SOURCESTATISTICS instead');
+assert(isnumeric(design), 'this function requires numeric data as input, you probably want to use FT_TIMELOCKSTATISTICS, FT_FREQSTATISTICS or FT_SOURCESTATISTICS instead');
+
+% flip dimensions such that the number of trials comes first
+dat = dat';
+
+% if cfg.dim has two entries which are non-singleton then the data has been 
+% 3D before. We must reshape it from 2D to 3D to run it in MVPA-Light
+data_is_3D = (numel(cfg.dim) > 1 && all(cfg.dim > 1));  % checks whether data was 3D before being reshaped
+if data_is_3D
+    dat = reshape(dat, size(dat,1), cfg.dim(1), cfg.dim(2));
+end
+
+y = cfg.design;
+
+%% build up config struct for MVPA-Light
+mvcfg = keepfields(cfg, {'balance','replace','normalise', ...
+                         'cv','k','repeat','p','stratify',...
+                         'classifier','param','size','nb'});
+
+mvcfg.metric          = ft_getopt(cfg, 'metric','accuracy');
+mvcfg.feedback        = ft_getopt(cfg, 'feedback','yes');
+
+% cfg: set defaults
+cfg.searchlight     = ft_getopt(cfg, 'searchlight','no');
+cfg.timextime       = ft_getopt(cfg, 'timextime','no');
+cfg.std             = ft_getopt(cfg, 'std', 'no');
+
+% translate parameter value into MVPA-Light notation [0 or 1]
+if ischar(mvcfg.feedback)
+    mvcfg.feedback = strcmp(mvcfg.feedback,'yes');
+end
+
+%% perform sanity checks on parameters
+if strcmp(cfg.timextime,'yes') && strcmp(cfg.searchlight,'yes')
+    ft_error('you should not set timextime = ''yes'' and searchlight = ''yes'' simultaneously')
+end
+
+% timextime = 'yes' but data is not 3D we should change timextime to 'no'
+if strcmp(cfg.timextime,'yes') && ~data_is_3D 
+    ft_warning('timextime = ''yes'' but data has no time dimension, setting timextime = ''no''');
+    cfg.timextime = 'no';
+end
+
+%% Call MVPA-Light 
+if strcmp(cfg.searchlight, 'yes')
+    % --- searchlight analysis ---
+    [perf, result] = mv_searchlight(mvcfg, dat, y);
+    
+elseif strcmp(cfg.timextime, 'yes')
+    % --- time x time generalisation ---
+    [perf, result] = mv_classify_timextime(mvcfg, dat, y);
+    
+elseif data_is_3D
+    % --- classification across time ---
+    [perf, result] = mv_classify_across_time(mvcfg, dat, y);
+
+else
+    % --- data has no time dimension, perform only cross-validation ---
+    [perf, result] = mv_crossvalidate(mvcfg, dat, y);
+end
+
+%% setup stat struct
+stat = [];
```

Comment:
We've partly had this conversation over email. Regarding the shape of the stats returned you said : 

>  I don’t think that this is problematic, because the reshape is only done if the number of elements in the field are equal to the number of data points in the input (without the observations obviously). That is, if I were you, I’d provide the correctly reshaped data already in the output from the mvpa wrapper function, which will then be passed on unchanged to the output of ft_timelockanalysis. In this case, you can make the data have an arbitrary ‘shape’, depending out the output metric.

So input and output data shapes will not match in some cases but this shouldn't be a problem as far as ft_timelock is concerned. I haven't tested freq/sourcestatistics yet (and I haven't written a tutorial on this either) it's something I'd prefer to look into in the future.

Reply:
indeed you're right. I was just asking (again). I'd be fine proceeding as is, and make improvements along the way.

Reply:
adding to this: if for instance the spatial dimension is reduced somehow we could consider outputting to the caller function an updated label field, and perhaps even an updated dim in the cfg, so that the reshape will take place accordingly. Just to keep in mind.

Reply:
Just note you can have multiple metrics with different shapes (eg 'accuracy' is a scalar, 'confusion' is a matrix). You could get it roughly right, for instance for time x time classification the resulting dim could be coded as time_time_metric but then metric could be 1D (for accuracy) or 2D (for confusion). Not sure if this makes sense though.

Reply:
Shall we revisit the dim/label question some other time? We'd either need different dim/label fields for different metrics or some kind of agreement that the metric is allowed to be multidimensional, at least I don't see an obvious elegant solution right now...